From 0d9658716ba43db6635ebb6637532cf6d4ee77df Mon Sep 17 00:00:00 2001
From: Radhey Shyam Pandey <radhey.shyam.pandey@xilinx.com>
Date: Tue, 1 Dec 2015 15:50:17 +0530
Subject: [PATCH 0205/1851] media: xilinx: Fix TPG incorrect pattern on
 resolution switch

commit 18abeeec0d29215cf71d2939ab6797aae52c1696 from
https://github.com/Xilinx/linux-xlnx.git

There is an known issue in TPG v7.0  that on resolution change
TPG generates incorrect pattern i.e hor/ver offset is added.
As a workaround assert hard reset on video pipeline stop.

Signed-off-by: Radhey Shyam Pandey <radhey.shyam.pandey@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
Tested-by: Christian Kohn <christian.kohn@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-tpg.c | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/drivers/media/platform/xilinx/xilinx-tpg.c b/drivers/media/platform/xilinx/xilinx-tpg.c
index 9710aa3d82a7..c97c38986fc8 100644
--- a/drivers/media/platform/xilinx/xilinx-tpg.c
+++ b/drivers/media/platform/xilinx/xilinx-tpg.c
@@ -208,11 +208,19 @@ static int xtpg_s_stream(struct v4l2_subdev *subdev, int enable)
 	u32 bayer_phase;
 
 	if (!enable) {
-		if (xtpg->is_hls)
-			/* Disable auto-start */
-			xvip_write(&xtpg->xvip, XVIP_CTRL_CONTROL, 0x0);
-		else
+		if (!xtpg->is_hls) {
 			xvip_stop(&xtpg->xvip);
+		} else {
+			/*
+			 * There is an known issue in TPG v7.0 that on
+			 * resolution change it doesn't generates pattern
+			 * correctly i.e some hor/ver offset is added.
+			 * As a workaround issue reset on stop.
+			 */
+			gpiod_set_value_cansleep(xtpg->rst_gpio, 0x1);
+			gpiod_set_value_cansleep(xtpg->rst_gpio, 0x0);
+			v4l2_ctrl_handler_setup(&xtpg->ctrl_handler);
+		}
 
 		if (xtpg->vtc)
 			xvtc_generator_stop(xtpg->vtc);
-- 
2.31.1

